File: native/android/voice/sakha/SakhaSseClient.kt
/**
 * Sakha SSE Client
 *
 * Streams the backend `/api/voice-companion/message` endpoint via OkHttp +
 * a hand-rolled SSE parser. Why hand-rolled and not okhttp-sse? okhttp-sse
 * uses the EventSource RFC, which expects GETs. Our endpoint is POST + SSE
 * (the body carries the prompt), which EventSource cannot speak.
 *
 * The client emits a typed [SakhaStreamEvent] flow:
 *  - Token  — incremental text from the model
 *  - Verse  — verse citation metadata
 *  - Engine — engine-determined mood/intensity
 *  - Audio  — base64 audio chunk if the backend pre-synthesised TTS
 *  - Done   — stream finished cleanly
 *  - Error  — terminal error
 *
 * Behaviour:
 *  - One outstanding request per [SakhaSseClient] instance. [cancel] aborts.
 *  - Bearer token attached if the config supplies one.
 *  - On 401, attempts ONE silent refresh via the supplied [refreshToken] hook.
 *  - On 429, surfaces [SakhaVoiceError.QuotaExceeded].
 *  - On 5xx, retries with exponential backoff up to [SakhaVoiceConfig.maxRequestRetries].
 *  - Does not parse the model output for pause markers — that's the parser's
 *    job downstream.
 */

package com.mindvibe.kiaan.voice.sakha

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.channels.ProducerScope
import kotlinx.coroutines.channels.awaitClose
import kotlinx.coroutines.delay
import kotlinx.coroutines.ensureActive
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.callbackFlow
import kotlinx.coroutines.flow.flowOn
import kotlinx.coroutines.suspendCancellableCoroutine
import okhttp3.Call
import okhttp3.Callback
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import okhttp3.Response
import org.json.JSONObject
import java.io.BufferedReader
import java.io.IOException
import java.util.concurrent.TimeUnit

sealed class SakhaStreamEvent {
    data class Token(val text: String) : SakhaStreamEvent()
    data class Engine(val engine: SakhaEngine, val mood: SakhaMood, val intensity: Int) : SakhaStreamEvent()
    data class Verse(val reference: String, val sanskrit: String?, val translation: String?) : SakhaStreamEvent()
    data class Audio(val base64: String, val mimeType: String) : SakhaStreamEvent()
    data class Session(val sessionId: String) : SakhaStreamEvent()
    object Done : SakhaStreamEvent()
    data class Error(val error: SakhaVoiceError) : SakhaStreamEvent()
}

class SakhaSseClient(
    private val config: SakhaVoiceConfig,
    private val refreshToken: suspend () -> String? = { null },
) {

    private val httpClient: OkHttpClient = OkHttpClient.Builder()
        .connectTimeout(config.requestTimeoutMs, TimeUnit.MILLISECONDS)
        .readTimeout(0, TimeUnit.MILLISECONDS) // SSE: never read-timeout
        .writeTimeout(config.requestTimeoutMs, TimeUnit.MILLISECONDS)
        .retryOnConnectionFailure(true)
        .build()

    @Volatile
    private var inflightCall: Call? = null

    /**
     * Open a Sakha streaming turn. The flow completes when the backend emits
     * [SakhaStreamEvent.Done] or a terminal error.
     */
    fun stream(
        userText: String,
        sessionId: String?,
        mood: SakhaMood? = null,
    ): Flow<SakhaStreamEvent> = callbackFlow {
        runRequest(userText, sessionId, mood)
        awaitClose {
            inflightCall?.cancel()
            inflightCall = null
        }
    }.flowOn(Dispatchers.IO)

    /** Cancel any in-flight request. Safe to call from any thread. */
    fun cancel() {
        inflightCall?.cancel()
        inflightCall = null
    }

    // ------------------------------------------------------------------------
    // Request driver
    // ------------------------------------------------------------------------

    /**
     * Drives the request lifecycle from inside the callbackFlow's
     * ProducerScope. We need the explicit receiver so [trySend] / [close]
     * resolve to the channel.
     */
    private suspend fun ProducerScope<SakhaStreamEvent>.runRequest(
        userText: String,
        sessionId: String?,
        mood: SakhaMood?,
    ) {
        var attempt = 0
        var token = config.authTokenProvider.invoke()
        var hasRetriedAuth = false

        while (true) {
            attempt++
            val url = "${config.backendBaseUrl.trimEnd('/')}${config.voiceCompanionPathPrefix}/message"
            val body = buildRequestBody(userText, sessionId, mood)

            val builder = Request.Builder()
                .url(url)
                .post(body.toRequestBody(JSON_MEDIA_TYPE))
                .header("Accept", "text/event-stream")
                .header("X-Client", "kiaanverse-android-sakha-voice")
                .header("X-Voice-Mode", "sakha")
            token?.let { builder.header("Authorization", "Bearer $it") }

            val call = httpClient.newCall(builder.build())
            inflightCall = call

            val outcome = try {
                executeOnce(call)
            } catch (io: IOException) {
                Outcome.NetworkFailure(io)
            } catch (t: Throwable) {
                Outcome.Unknown(t.message ?: "unknown")
            }

            when (outcome) {
                is Outcome.AuthExpired -> {
                    if (!hasRetriedAuth) {
                        hasRetriedAuth = true
                        val refreshed = refreshToken()
                        if (refreshed != null) {
                            token = refreshed
                            continue
                        }
                    }
                    trySend(SakhaStreamEvent.Error(SakhaVoiceError.AuthRequired))
                    close()
                    return
                }
                is Outcome.Quota -> {
                    trySend(SakhaStreamEvent.Error(SakhaVoiceError.QuotaExceeded))
                    close()
                    return
                }
                is Outcome.Server -> {
                    if (attempt <= config.maxRequestRetries) {
                        delay(backoff(attempt))
                        continue
                    }
                    trySend(SakhaStreamEvent.Error(SakhaVoiceError.HttpError(outcome.status, "Server error")))
                    close()
                    return
                }
                is Outcome.HttpFailure -> {
                    trySend(SakhaStreamEvent.Error(SakhaVoiceError.HttpError(outcome.status, outcome.detail ?: "Request failed")))
                    close()
                    return
                }
                is Outcome.NetworkFailure -> {
                    if (attempt <= config.maxRequestRetries) {
                        delay(backoff(attempt))
                        continue
                    }
                    trySend(SakhaStreamEvent.Error(SakhaVoiceError.NetworkError(outcome.cause.message ?: "io")))
                    close()
                    return
                }
                is Outcome.Unknown -> {
                    trySend(SakhaStreamEvent.Error(SakhaVoiceError.Unknown(outcome.detail)))
                    close()
                    return
                }
                is Outcome.Streamed -> {
                    consumeStream(outcome.response)
                    trySend(SakhaStreamEvent.Done)
                    close()
                    return
                }
            }
        }
    }

    private suspend fun executeOnce(call: Call): Outcome {
        val response = call.executeSuspending()
        val status = response.code

        if (status == 401 || status == 403) {
            response.closeQuietly()
            return Outcome.AuthExpired
        }
        if (status == 429) {
            response.closeQuietly()
            return Outcome.Quota
        }
        if (status >= 500) {
            response.closeQuietly()
            return Outcome.Server(status)
        }
        if (!response.isSuccessful) {
            val detail = try { response.body?.string()?.take(200) } catch (_: Exception) { null }
            response.closeQuietly()
            return Outcome.HttpFailure(status, detail)
        }
        return Outcome.Streamed(response)
    }

    private fun backoff(attempt: Int): Long = 500L * (1L shl (attempt - 1).coerceAtMost(4))

    // ------------------------------------------------------------------------
    // SSE consumption
    // ------------------------------------------------------------------------

    private suspend fun ProducerScope<SakhaStreamEvent>.consumeStream(response: Response) {
        val source = response.body?.charStream() ?: return
        val reader = BufferedReader(source)
        try {
            while (true) {
                ensureActive()
                val line = reader.readLine() ?: break
                val trimmed = line.trim()
                if (trimmed.isEmpty()) continue
                if (!trimmed.startsWith("data:")) continue
                val payload = trimmed.removePrefix("data:").trim()
                if (payload.isEmpty()) continue
                val event = parseFrame(payload) ?: continue
                trySend(event)
                if (event is SakhaStreamEvent.Done) return
            }
        } finally {
            try { reader.close() } catch (_: Exception) {}
            response.closeQuietly()
        }
    }

    /**
     * Parse a single SSE frame payload. The backend uses a few overlapping
     * shapes (see backend/routes/kiaan_voice_companion.py); we accept all of
     * them and normalise to [SakhaStreamEvent].
     */
    private fun parseFrame(payload: String): SakhaStreamEvent? {
        if (payload == "[DONE]") return SakhaStreamEvent.Done

        val json = try {
            JSONObject(payload)
        } catch (_: Exception) {
            // Legacy plain-text frame — treat as a token.
            return SakhaStreamEvent.Token(payload)
        }

        when (json.optString("type")) {
            "token" -> return SakhaStreamEvent.Token(json.optString("text", ""))
            "tts_chunk" -> return SakhaStreamEvent.Token(json.optString("text", ""))
            "voice_emotion" -> {
                val emotion = json.optString("emotion", "neutral")
                return SakhaStreamEvent.Engine(
                    engine = config.defaultEngine,
                    mood = SakhaMood.fromWire(emotion),
                    intensity = json.optInt("intensity", 5),
                )
            }
            "audio" -> return SakhaStreamEvent.Audio(
                base64 = json.optString("data", ""),
                mimeType = json.optString("mime", "audio/wav"),
            )
            "verse" -> return SakhaStreamEvent.Verse(
                reference = json.optString("ref", ""),
                sanskrit = json.optString("sanskrit").ifEmpty { null },
                translation = json.optString("translation").ifEmpty { null },
            )
            "session" -> return SakhaStreamEvent.Session(json.optString("session_id", ""))
            "done" -> return SakhaStreamEvent.Done
            "error" -> {
                val code = json.optString("code", "")
                val msg = json.optString("message", "Unknown")
                return SakhaStreamEvent.Error(when (code) {
                    "quota_exceeded" -> SakhaVoiceError.QuotaExceeded
                    "auth_required" -> SakhaVoiceError.AuthRequired
                    else -> SakhaVoiceError.Unknown(msg)
                })
            }
        }

        // Legacy `{"word": "...", "done": false}` shape used by /chat/message/stream
        if (json.has("word")) return SakhaStreamEvent.Token(json.optString("word", ""))
        if (json.optBoolean("done", false)) return SakhaStreamEvent.Done
        if (json.has("session_id")) return SakhaStreamEvent.Session(json.optString("session_id", ""))
        return null
    }

    // ------------------------------------------------------------------------
    // Request body
    // ------------------------------------------------------------------------

    private fun buildRequestBody(userText: String, sessionId: String?, mood: SakhaMood?): String {
        val payload = JSONObject().apply {
            put("message", userText)
            put("language", config.language.wire)
            put("voice_id", config.sakhaVoiceId)
            put("content_type", "voice")
            put("voice_mode", "sakha")
            put("stream", true)
            sessionId?.let { put("session_id", it) }
            mood?.let { put("mood_hint", it.wire) }
        }
        return payload.toString()
    }

    // ------------------------------------------------------------------------
    // OkHttp helpers
    // ------------------------------------------------------------------------

    private suspend fun Call.executeSuspending(): Response =
        suspendCancellableCoroutine { cont ->
            enqueue(object : Callback {
                override fun onFailure(call: Call, e: IOException) {
                    if (cont.isActive) cont.resumeWith(Result.failure(e))
                }

                override fun onResponse(call: Call, response: Response) {
                    if (cont.isActive) cont.resumeWith(Result.success(response))
                }
            })
            cont.invokeOnCancellation { runCatching { cancel() } }
        }

    private fun Response.closeQuietly() {
        try { close() } catch (_: Exception) {}
    }

    // ------------------------------------------------------------------------
    // Outcome (internal)
    // ------------------------------------------------------------------------

    private sealed class Outcome {
        object AuthExpired : Outcome()
        object Quota : Outcome()
        data class Server(val status: Int) : Outcome()
        data class HttpFailure(val status: Int, val detail: String?) : Outcome()
        data class NetworkFailure(val cause: IOException) : Outcome()
        data class Unknown(val detail: String) : Outcome()
        data class Streamed(val response: Response) : Outcome()
    }

    companion object {
        private val JSON_MEDIA_TYPE = "application/json; charset=utf-8".toMediaType()
    }
}

File: kiaanverse-mobile/apps/mobile/native/ios/Sources/Sakha/SakhaSseClient.swift
/**
 * Sakha SSE Client (iOS port)
 *
 * Direct port of apps/mobile/native/android/.../SakhaSseClient.kt. Uses
 * URLSessionDataTask with a delegate to receive incremental chunks of
 * the POST→SSE response from `/api/voice-companion/sakha/stream`. The
 * backend speaks SSE-over-POST (not RFC EventSource) because the prompt
 * lives in the body — same constraint as Android.
 *
 * Emits typed `SakhaStreamEvent` values via a callback. Single in-flight
 * request per instance; `cancel()` aborts.
 *
 * Auth + retry policy matches Android exactly:
 *   - 401/403: one silent refreshToken attempt, then SakhaVoiceError.authRequired
 *   - 429:     SakhaVoiceError.quotaExceeded (no retry)
 *   - 5xx:     exponential backoff, up to config.maxRequestRetries
 *   - I/O:     same backoff, same cap
 */

import Foundation

public enum SakhaStreamEvent {
    case token(text: String)
    case engine(engine: SakhaEngine, mood: SakhaMood, intensity: Int)
    case verse(reference: String, sanskrit: String?, translation: String?)
    case audio(base64: String, mimeType: String)
    case session(sessionId: String)
    case done
    case error(SakhaVoiceError)
}

public final class SakhaSseClient {

    private let config: SakhaVoiceConfig
    /// Async closure that returns a refreshed bearer token, or nil to give up.
    private let refreshToken: () async -> String?
    /// Async closure that returns the current bearer token (or nil).
    private let authTokenProvider: () async -> String?

    private let session: URLSession
    private let delegate: SseDelegate
    private var task: URLSessionDataTask?
    private var isCancelled = false

    public init(
        config: SakhaVoiceConfig,
        authTokenProvider: @escaping () async -> String?,
        refreshToken: @escaping () async -> String? = { nil }
    ) {
        self.config = config
        self.authTokenProvider = authTokenProvider
        self.refreshToken = refreshToken

        let sessionConfig = URLSessionConfiguration.default
        sessionConfig.timeoutIntervalForRequest = TimeInterval(config.requestTimeoutMs) / 1000.0
        // SSE is open-ended — we never want a "read timeout" while the
        // server is happily streaming tokens.
        sessionConfig.timeoutIntervalForResource = .greatestFiniteMagnitude

        let delegate = SseDelegate()
        self.delegate = delegate
        self.session = URLSession(configuration: sessionConfig, delegate: delegate, delegateQueue: nil)
    }

    /// Open a Sakha streaming turn. The handler is invoked for each parsed
    /// SakhaStreamEvent in arrival order. After `.done` or `.error(...)`,
    /// the handler is not invoked again.
    public func stream(
        userText: String,
        sessionId: String?,
        mood: SakhaMood?,
        turnCount: Int,
        onEvent: @escaping (SakhaStreamEvent) -> Void
    ) {
        isCancelled = false
        Task { await self.runRequest(userText: userText, sessionId: sessionId, mood: mood, turnCount: turnCount, onEvent: onEvent) }
    }

    /// Cancel any in-flight request.
    public func cancel() {
        isCancelled = true
        task?.cancel()
        task = nil
    }

    // MARK: - Request driver

    private func runRequest(
        userText: String,
        sessionId: String?,
        mood: SakhaMood?,
        turnCount: Int,
        onEvent: @escaping (SakhaStreamEvent) -> Void
    ) async {
        var attempt = 0
        var token = await authTokenProvider()
        var hasRetriedAuth = false

        while !isCancelled {
            attempt += 1

            let url = "\(trimSlash(config.backendBaseUrl))\(config.voiceCompanionPathPrefix)/sakha/stream"
            guard let endpoint = URL(string: url) else {
                onEvent(.error(.unknown("invalid url: \(url)")))
                return
            }

            var request = URLRequest(url: endpoint)
            request.httpMethod = "POST"
            request.setValue("application/json; charset=utf-8", forHTTPHeaderField: "Content-Type")
            request.setValue("text/event-stream", forHTTPHeaderField: "Accept")
            request.setValue("kiaanverse-ios-sakha-voice", forHTTPHeaderField: "X-Client")
            request.setValue("sakha", forHTTPHeaderField: "X-Voice-Mode")
            if let token = token, !token.isEmpty {
                request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
            }
            request.httpBody = buildRequestBody(userText: userText, sessionId: sessionId, mood: mood, turnCount: turnCount)

            // Wire delegate handlers for this attempt.
            delegate.reset(onLine: { [weak self] payload in
                guard let self = self, !self.isCancelled else { return }
                if let event = self.parseFrame(payload) {
                    onEvent(event)
                    if case .done = event {
                        self.task?.cancel()
                    }
                }
            })

            let outcome: AttemptOutcome = await withCheckedContinuation { continuation in
                let dataTask = session.dataTask(with: request)
                self.task = dataTask
                delegate.onCompletion = { status, error in
                    if let error = error as NSError?, error.domain == NSURLErrorDomain, error.code == NSURLErrorCancelled {
                        continuation.resume(returning: .cancelled)
                        return
                    }
                    if let error = error {
                        continuation.resume(returning: .networkFailure(error.localizedDescription))
                        return
                    }
                    if status == 401 || status == 403 {
                        continuation.resume(returning: .authExpired)
                        return
                    }
                    if status == 429 {
                        continuation.resume(returning: .quota)
                        return
                    }
                    if status >= 500 {
                        continuation.resume(returning: .server(status))
                        return
                    }
                    if status >= 400 {
                        continuation.resume(returning: .httpFailure(status))
                        return
                    }
                    continuation.resume(returning: .streamed)
                }
                dataTask.resume()
            }

            switch outcome {
            case .streamed:
                // delegate already streamed events including .done if backend sent it.
                onEvent(.done)
                return
            case .cancelled:
                return
            case .authExpired:
                if !hasRetriedAuth {
                    hasRetriedAuth = true
                    if let refreshed = await refreshToken() {
                        token = refreshed
                        continue
                    }
                }
                onEvent(.error(.authRequired))
                return
            case .quota:
                onEvent(.error(.quotaExceeded))
                return
            case .server(let status):
                if attempt <= config.maxRequestRetries {
                    try? await Task.sleep(nanoseconds: backoffNanos(attempt: attempt))
                    continue
                }
                onEvent(.error(.httpError(status, "Server error")))
                return
            case .httpFailure(let status):
                onEvent(.error(.httpError(status, "Request failed")))
                return
            case .networkFailure(let detail):
                if attempt <= config.maxRequestRetries {
                    try? await Task.sleep(nanoseconds: backoffNanos(attempt: attempt))
                    continue
                }
                onEvent(.error(.networkError(detail)))
                return
            }
        }
    }

    private func backoffNanos(attempt: Int) -> UInt64 {
        let shifted = min(attempt - 1, 4)
        let ms: UInt64 = 500 * (1 << shifted)
        return ms * 1_000_000
    }

    private func trimSlash(_ s: String) -> String {
        if s.hasSuffix("/") { return String(s.dropLast()) }
        return s
    }

    private func buildRequestBody(userText: String, sessionId: String?, mood: SakhaMood?, turnCount: Int) -> Data {
        var payload: [String: Any] = [
            "message": userText,
            "language": config.language.wire,
            "turn_count": max(turnCount, 1),
        ]
        if let sessionId = sessionId { payload["session_id"] = sessionId }
        if let mood = mood { payload["mood_hint"] = mood.wire }
        return (try? JSONSerialization.data(withJSONObject: payload, options: [])) ?? Data()
    }

    // MARK: - SSE frame parsing

    /// Parse a single SSE frame payload (the part after `data:`). Returns
    /// nil for frames the consumer should silently skip (e.g. tts_chunk
    /// repeats — already segmented by SakhaPauseParser).
    private func parseFrame(_ payload: String) -> SakhaStreamEvent? {
        if payload == "[DONE]" { return .done }

        guard let data = payload.data(using: .utf8),
              let json = try? JSONSerialization.jsonObject(with: data, options: []) as? [String: Any]
        else {
            // Legacy plain-text frame — treat as a token.
            return .token(text: payload)
        }

        let type = (json["type"] as? String) ?? ""

        switch type {
        case "token":
            return .token(text: (json["text"] as? String) ?? "")
        case "tts_chunk":
            // Parser-side dedupe: tts_chunk repeats accumulated tokens that
            // we've already parsed via "token" frames. Ignore.
            return nil
        case "engine":
            return .engine(
                engine: SakhaEngine.fromWire(json["engine"] as? String),
                mood: SakhaMood.fromWire(json["mood"] as? String),
                intensity: (json["intensity"] as? Int) ?? 5
            )
        case "voice_emotion":
            return .engine(
                engine: config.defaultEngine,
                mood: SakhaMood.fromWire(json["emotion"] as? String),
                intensity: (json["intensity"] as? Int) ?? 5
            )
        case "audio":
            return .audio(
                base64: (json["data"] as? String) ?? "",
                mimeType: (json["mime"] as? String) ?? "audio/wav"
            )
        case "verse":
            let sanskrit = (json["sanskrit"] as? String).flatMap { $0.isEmpty ? nil : $0 }
            let translation = (json["translation"] as? String).flatMap { $0.isEmpty ? nil : $0 }
            return .verse(
                reference: (json["ref"] as? String) ?? "",
                sanskrit: sanskrit,
                translation: translation
            )
        case "session":
            return .session(sessionId: (json["session_id"] as? String) ?? "")
        case "done":
            return .done
        case "error":
            let code = (json["code"] as? String) ?? ""
            let msg = (json["message"] as? String) ?? "Unknown"
            switch code {
            case "quota_exceeded": return .error(.quotaExceeded)
            case "auth_required":  return .error(.authRequired)
            default:               return .error(.unknown(msg))
            }
        default:
            // Legacy `{"word": "..."}` and `{"done": true}` shapes from /chat/message/stream.
            if let word = json["word"] as? String { return .token(text: word) }
            if let done = json["done"] as? Bool, done { return .done }
            if let session = json["session_id"] as? String { return .session(sessionId: session) }
            return nil
        }
    }

    // MARK: - Outcome

    private enum AttemptOutcome {
        case streamed
        case cancelled
        case authExpired
        case quota
        case server(Int)
        case httpFailure(Int)
        case networkFailure(String)
    }
}

// MARK: - URLSessionDataDelegate

/// Delegate buffers incoming bytes and emits one SSE `data:` payload at
/// a time to its `onLine` handler. Lifecycle: configured per-attempt by
/// the client via `reset(onLine:)`. `onCompletion` fires once when the
/// task ends with the final HTTP status (or 0 + error).
private final class SseDelegate: NSObject, URLSessionDataDelegate {

    var onCompletion: ((Int, Error?) -> Void)?
    private var onLine: ((String) -> Void)?
    private var buffer = Data()
    private var status: Int = 0

    func reset(onLine: @escaping (String) -> Void) {
        self.onLine = onLine
        self.buffer = Data()
        self.status = 0
    }

    func urlSession(_ session: URLSession,
                    dataTask: URLSessionDataTask,
                    didReceive response: URLResponse,
                    completionHandler: @escaping (URLSession.ResponseDisposition) -> Void) {
        if let http = response as? HTTPURLResponse {
            status = http.statusCode
        }
        completionHandler(.allow)
    }

    func urlSession(_ session: URLSession, dataTask: URLSessionDataTask, didReceive data: Data) {
        buffer.append(data)
        flushLines()
    }

    func urlSession(_ session: URLSession, task: URLSessionTask, didCompleteWithError error: Error?) {
        // Final flush in case the server closed without a trailing newline.
        flushLines()
        onCompletion?(status, error)
    }

    /// Walks `buffer`, emits each newline-terminated `data:` payload, and
    /// keeps the trailing partial line for the next chunk.
    private func flushLines() {
        while let nlRange = buffer.range(of: Data([0x0A])) {
            let lineData = buffer.subdata(in: 0..<nlRange.lowerBound)
            buffer.removeSubrange(0..<nlRange.upperBound)

            guard let line = String(data: lineData, encoding: .utf8) else { continue }
            let trimmed = line.trimmingCharacters(in: .whitespacesAndNewlines)
            if trimmed.isEmpty { continue }
            guard trimmed.hasPrefix("data:") else { continue }
            let payload = String(trimmed.dropFirst("data:".count)).trimmingCharacters(in: .whitespaces)
            if payload.isEmpty { continue }
            onLine?(payload)
        }
    }
}
